### PR TITLE
Dockerfile tests

### PR DESCRIPTION
**Pull request by robertbeal**
I've started a small suite of tests for the Dockerfile, mainly to ensure we don't break anything and allows us to test-drive any changes we make

### CODE REVIEW

**Comment by robertbeal** [comment]
(note - not sure how easily it'll integrate with travis, may need some tweaking)

**Comment by robertbeal** [comment]
Will need to squash the commits as it's taken a bit of tweaking to get it passing on travis

File: test.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    docker_id = subprocess.check_output(['docker', 'run', '-d', 'radicale']).decode().strip()
+    
+    yield testinfra.get_host("docker://" + docker_id)
+    
+    # teardown
+    subprocess.check_call(['docker', 'rm', '-f', docker_id])
+
+def test_system(host):
+    assert host.system_info.distribution == 'alpine'
+    assert host.system_info.release == '3.7.0'
```
Comment (tomsquest) [inline test.py]: 3.8.0 now, no ?

File: Dockerfile [hunk 2]
```diff
@@ -31,5 +32,6 @@ VOLUME /config /data
 EXPOSE 5232
 
 COPY docker-entrypoint.sh /usr/local/bin
+RUN chmod 555 /usr/local/bin/docker-entrypoint.sh
```
Comment (tomsquest) [inline Dockerfile]: why is this needed (because the script is already executable in the repo, no?)

File: README.md [hunk 3]
```diff
@@ -105,6 +105,16 @@ To customize Radicale configuration, either:
 
 Then puts these two files in a directory and use the config volume `-v /my_custom_config_directory:/config` when running the container.
 
+## Contributing
+
+First install the test dependencies
+
+`pip install --user -r requirements.txt`
```
Comment (tomsquest) [inline README.md]: As the requirements file is only for testing, what do you think of naming it `requirements-test.txt` (eg. https://github.com/uber/tchannel-python/blob/master/requirements-test.txt ) ?

File: test.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import os
+import pytest
+import subprocess
+import testinfra
+
+@pytest.fixture(scope='session')
+def host(request):
+    subprocess.check_call(['docker', 'build', '-t', 'radicale', '.'])
```
Comment (tomsquest) [inline test.py]: The image name can be skipped, which can then avoid a conflict with a local `radicale` image, no?

In bash:
```
IMAGE_ID=$(docker build -q .)
docker run -it --rm "$IMAGE_ID"
```

File: test.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    docker_id = subprocess.check_output(['docker', 'run', '-d', 'radicale']).decode().strip()
+    
+    yield testinfra.get_host("docker://" + docker_id)
+    
+    # teardown
+    subprocess.check_call(['docker', 'rm', '-f', docker_id])
+
+def test_system(host):
```
Comment (tomsquest) [inline test.py]: I would skip this test

File: test.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    assert host.system_info.distribution == 'alpine'
+    assert host.system_info.release == '3.7.0'
+
+def test_entrypoint(host):
```
Comment (tomsquest) [inline test.py]: Same here, I would skip this test (see main comment)

File: test.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    entrypoint = '/usr/local/bin/docker-entrypoint.sh'
+    assert host.file(entrypoint).exists
+    assert oct(host.file(entrypoint).mode) == '0o555'
+
+def test_process(host):
+    process = host.process.get(comm='radicale')
+    assert process.pid == 1
+    assert process.user == 'radicale'
```
Comment (tomsquest) [inline test.py]: :+1:

File: test.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    assert '/usr/bin/radicale --config /config/config' in process.args
+
+def test_port(host):
+    assert host.socket("tcp://0.0.0.0:5232").is_listening
```
Comment (tomsquest) [inline test.py]: This test case could set a custom port and try if radicale is bound to it

File: test.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+def test_radicale_version(host):
+    assert host.check_output("/usr/bin/pip3 --disable-pip-version-check show radicale | grep Version | awk -F ' ' '{print $2}' | tr -d '\n'") == os.environ.get('VERSION','2.1.9')
```
Comment (tomsquest) [inline test.py]: can't radicale be asked its running version using curl?

File: test.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+def test_radicale_user(host):
```
Comment (tomsquest) [inline test.py]: would skip this test

File: test.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    user = 'radicale'
+    assert host.user(user).uid == 2999
```
Comment (tomsquest) [inline test.py]: :+1:

File: Dockerfile (hunk 2 above)
Comment (robertbeal) [reply Dockerfile]: Yeah can definitely black box test radicale. 

The reason for the permission check is that by default (when using `COPY`) the file has write permissions (755), although oddly it differed between running it locally and on travis (775 vs 755). So the thinking was to be explicit and set read/execute permissions only as that's all that is needed.

Granted it's moot if running your container as `--read-only` though

File: test.py (hunk 4 above)
Comment (robertbeal) [reply test.py]: Ah well spotted. It can either be skipped, or in test I could call it something like `image-under-test`. 

Actually in order to tests to be self contained, probably best to do the latter otherwise you'd have to build radicale locally each time prior to running tests.

File: test.py (hunk 5 above)
Comment (robertbeal) [reply test.py]: Yeah, added it because I could, but it doesn't offer much value

File: test.py (hunk 6 above)
Comment (robertbeal) [reply test.py]: I'll drop it. As you mention the entrypoint is just a means to running radicale, so from a black box perspective it doesn't really matter that it exists, just that radicale starts up

File: test.py (hunk 5 above)
Comment (robertbeal) [reply test.py]: Ah, actually one reason it does add value is in the build.sh. It'd be easy to have one of the ARCH builds on a different version of alpine by mistake. This would ensure that doesn't happen

File: test.py (hunk 9 above)
Comment (robertbeal) [reply test.py]: Ah yes, `radicale --version`, much cleaner

File: test.py (hunk 8 above)
Comment (robertbeal) [reply test.py]: It would mean changing the default config prior to building the container, which would then differ to the actual image we build. 

I like the concept of testing a random port from a testing radicale perspective as that should work. Whereas this test it to check the image we build, runs by default on port 5232.

File: test.py (hunk 10 above)
Comment (robertbeal) [reply test.py]: Is it worth leaving in until we have black box tests in place? I've had a quick look but can't see a "version" or "status" like endpoint to curl against a running radicale

**Comment by tomsquest** [comment]
Hi @robertbeal ,

I have started to read and test your PR.
I am merging it.

Next, I may try to:
- switch to pipenv (which I use locally), hence dropping the requirements-test.txt and adding the Pipfile files. This would also require to update the travis.yml according to https://pipenv.readthedocs.io/en/latest/advanced/#travis-ci
- adding this test: changing the UID/GID, so testing that the radicale process is run by this user
